File: 03-week-3--additional-practice/00-bonus-1--balancing-parenetheses/javascript/balancing_parentheses.js
function balancingParentheses(string) {
  // type your code here
  let end = 0
  let front = 0

  for (let i = 0; i < string.length; ++i) {
    if (string[i] === '(') {
      ++openings
      continue
    }

    if (front > 0) {
      --front
    } else {
      ++end
    }
  }
  return end + front
}

}

if (require.main === module) {
  // add your own tests in here
  console.log("Expecting: 0");
  console.log(balancingParentheses('(()())'));

  console.log("");

  console.log("Expecting: 2");
  console.log(balancingParentheses('()))'));

  console.log("");

  console.log("Expecting: 1");
  console.log(balancingParentheses(')'));
}

module.exports = balancingParentheses;

// Please add your pseudocode to this file
// And a written explanation of your solution

File: 03-week-3--additional-practice/00-bonus-1--balancing-parenetheses/ruby/balancing_parentheses.rb
def balancing_parentheses(string)
  # type your code in here
  end_count = 0
  opening_count = 0
  string.chars.each do |char|
    if char == '('
      opening_count += 1
      next
    end

    if opening_count > 0 
      opening_count -= 1
    else
      end_count += 1
    end
  end

  end_count + opening_count
end

if __FILE__ == $PROGRAM_NAME
  puts "Expecting: 0"
  puts balancing_parentheses('(()())')

  puts

  puts "Expecting: 2"
  puts balancing_parentheses('()))')

  puts

  puts "Expecting: 1"
  puts balancing_parentheses(')')

  # Don't forget to add your own!
end

# Please add your pseudocode to this file
# And a written explanation of your solution
